### PR TITLE
refactor: new tdp client

### PR DESCRIPTION
**Pull request by jayzhang**

### CODE REVIEW

File: packages/fx-core/src/client/teamsDevPortalClient.ts [hunk 1]
```diff
@@ -0,0 +1,947 @@
...
+    }
+  }
+
+  setRegion(region: string) {
+    this.region = region;
+  }
+
+  async setRegionByToken(regionToken: string) {
+    this.region = undefined;
+    const requester = this.createRequesterWithToken(regionToken);
```
Comment (anchenyi) [inline packages/fx-core/src/client/teamsDevPortalClient.ts:103]: baseUrl should be authsvc? The createRequesterWithToken uses TDP baseUrl

File: packages/fx-core/src/client/teamsDevPortalClient.ts [hunk 2]
```diff
@@ -0,0 +1,947 @@
...
+          await new Promise((resolve) => setTimeout(resolve, 5000));
+        }
+      }
+    }
+  }
+}
+
+class TeamsDevPortalClient {
+  endpoint: string;
+  region?: string;
```
Comment (anchenyi) [inline packages/fx-core/src/client/teamsDevPortalClient.ts:84]: Can we call them `globalEndpoint `and `regionalEndpoint`? Normally, `region `should be a country code, not a regional url.

File: packages/fx-core/src/client/teamsDevPortalClient.ts [hunk 3]
```diff
@@ -0,0 +1,947 @@
...
+      }
+      const error = this.wrapException(e, APP_STUDIO_API_NAMES.CREATE_APP);
+      throw error;
+    }
+  }
+
+  @hooks([ErrorContextMW({ source: "Teams", component: "TeamsDevPortalClient" })])
+  async listApps(token: string): Promise<AppDefinition[]> {
+    if (!this.region) throw new Error("Failed to get region");
+    let requester: AxiosInstance;
```
Comment (anchenyi) [inline packages/fx-core/src/client/teamsDevPortalClient.ts:206]: `setErrorContext({ source: "Teams" });` is removed, is it expected?

File: packages/fx-core/src/client/teamsDevPortalClient.ts [hunk 4]
```diff
@@ -0,0 +1,947 @@
...
+    } catch (e) {
+      const error = this.wrapException(e, APP_STUDIO_API_NAMES.DELETE_APP);
+      throw error;
+    }
+    throw new Error("Cannot delete the app: " + teamsAppId);
+  }
+  @hooks([ErrorContextMW({ source: "Teams", component: "TeamsDevPortalClient" })])
+  async getApp(token: string, teamsAppId: string): Promise<AppDefinition> {
+    let requester: AxiosInstance;
+    for (const regional of [true, false]) {
```
Comment (anchenyi) [inline packages/fx-core/src/client/teamsDevPortalClient.ts:252]: Should we only call the regional endpoint?

File: packages/fx-core/src/client/teamsDevPortalClient.ts (hunk 3 above)
Comment (anchenyi) [reply packages/fx-core/src/client/teamsDevPortalClient.ts:206]: saw the hooks, can ignore this.

File: packages/fx-core/src/client/teamsDevPortalClient.ts [hunk 5]
```diff
@@ -0,0 +1,947 @@
...
+    }
+  }
+
+  @hooks([ErrorContextMW({ source: "Teams", component: "TeamsDevPortalClient" })])
+  async grantPermission(token: string, teamsAppId: string, newUser: AppUser): Promise<void> {
+    const app = await this.getApp(token, teamsAppId);
+    if (this.checkUser(app, newUser)) {
+      return;
+    }
+    app.userList?.push(newUser);
```
Comment (anchenyi) [inline packages/fx-core/src/client/teamsDevPortalClient.ts:521]: `const findUser = app.userList?.findIndex((user: AppUser) => user["aadId"] === newUser.aadId);
    if (findUser && findUser >= 0) {
      return;
    }`
is removed, expected?

Comment (anchenyi) [reply packages/fx-core/src/client/teamsDevPortalClient.ts:521]: checkUser already covered this, can ignore.

File: packages/fx-core/src/client/teamsDevPortalClient.ts (hunk 1 above)
Comment (jayzhang) [reply packages/fx-core/src/client/teamsDevPortalClient.ts]: Yes, I will fix.

**Review by anchenyi — APPROVED** [review]

**Review by LongOddCode — APPROVED** [review]

**Pull request merged**